mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/MapboxTelemetry.java

```diff
@@ -0,0 +1,573 @@
+import android.location.Location;
+import android.location.LocationManager;
+import android.os.BatteryManager;
+import android.os.Build;
+import android.os.Handler;
+import android.support.annotation.NonNull;
+import android.support.v4.content.LocalBroadcastManager;
+import android.text.TextUtils;
+import android.util.DisplayMetrics;
+import android.view.WindowManager;
+
+import com.mapbox.services.android.telemetry.backoff.ExponentialBackoff;
+import com.mapbox.services.android.telemetry.connectivity.ConnectivityReceiver;
+import com.mapbox.services.android.telemetry.constants.TelemetryConstants;
+import com.mapbox.services.android.telemetry.http.TelemetryClient;
+import com.mapbox.services.android.telemetry.location.AndroidLocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngineListener;
+import com.mapbox.services.android.telemetry.permissions.PermissionsManager;
+import com.mapbox.services.android.telemetry.service.TelemetryService;
+import com.mapbox.services.android.telemetry.utils.TelemetryUtils;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Hashtable;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.UUID;
+import java.util.Vector;
+
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.Response;
+import okhttp3.internal.Util;
+import timber.log.Timber;
+
+/**
+ * This is the entry point to manage Mapbox telemetry
+ */
+public class MapboxTelemetry implements Callback, LocationEngineListener {
+
+  private static MapboxTelemetry instance;
+
+  private boolean initialized = false;
+  private Context context = null;
+  private String accessToken = null;
+  private String mapboxSessionId = null;
+  private long mapboxSessionIdLastSet = 0;
+  private String mapboxVendorId = null;
+  private DisplayMetrics displayMetrics = null;
+  private Intent batteryStatus = null;
+  private TelemetryClient client = null;
+  private Vector<Hashtable<String, Object>> events = new Vector<>();
+  private Timer timer = null;
+  private LocationEngine locationEngine = null;
+  private boolean withShutDown = false;
+  private Boolean telemetryEnabled = null;
+
+  /**
+   * Private constructor for configuring the single instance per app.
+   */
+  private MapboxTelemetry() {
+  }
+
+  /**
+   * Primary access method (using singleton pattern)
+   *
+   * @return MapboxTelemetry
+   */
+  public static MapboxTelemetry getInstance() {
+    if (instance == null) {
+      instance = new MapboxTelemetry();
+    }
+
+    return instance;
+  }
+
+  /**
+   * Initialize MapboxTelemetry.
+   *
+   * @param context        The context associated with the application
+   * @param accessToken    The accessToken associated with the application
+   * @param locationEngine Initialize telemetry with a custom location engine
+   */
+  public void initialize(@NonNull Context context, @NonNull String accessToken,
```

Comment:
Some thoughts: I'm not a big supporter of the Singleton paradigm with a separate initialization method. I think this can be overcome by using a separate parameterized getInstance method and handling those parameters as static (see MapboxMap.java in GL SDK) or by extracting some logic and wrapping it around a factory or other design pattern.